src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/async_data_uploader.h"
+
+#include "archival/logger.h"
+#include "archival/types.h"
+#include "base/vlog.h"
+#include "model/record.h"
+#include "model/record_batch_types.h"
+#include "model/timeout_clock.h"
+#include "model/timestamp.h"
+#include "resource_mgmt/io_priority.h"
+#include "ssx/future-util.h"
+#include "storage/disk_log_impl.h"
+#include "storage/offset_to_filepos.h"
+#include "storage/parser.h"
+#include "storage/segment_appender_utils.h"
+#include "storage/segment_index.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/when_all.hh>
+
+#include <exception>
+#include <stdexcept>
+#include <system_error>
+
+namespace archival {
+
+class reader_ds : public ss::data_source_impl {
+    /// Consumer that reads data batches to the read buffer
+    /// of the data source.
+    class consumer {
+    public:
+        explicit consumer(reader_ds* ds, inclusive_offset_range range)
+          : _parent(ds)
+          , _range(range) {}
+
+        /// This consumer accepts all batches if there is enough space in the
+        /// read buffer.
+        ss::future<ss::stop_iteration> operator()(model::record_batch batch) {
+            vlog(
+              _parent->_ctxlog.debug,
+              "consume started, buffered {} bytes",
+              _parent->_buffer.size_bytes());
+            if (!_range.contains(batch.header())) {
+                vlog(
+                  _parent->_ctxlog.debug,
+                  "skip batch {}-{}",
+                  batch.base_offset(),
+                  batch.last_offset());
+
+                co_return ss::stop_iteration::no;
+            }
+            vlog(
+              _parent->_ctxlog.debug,
+              "consuming batch {}-{}",
+              batch.base_offset(),
+              batch.last_offset());
+            auto hdr_iobuf = storage::disk_header_to_iobuf(batch.header());
+            auto rec_iobuf = std::move(batch).release_data();
+            _parent->_buffer.append(std::move(hdr_iobuf));
+            _parent->_buffer.append(std::move(rec_iobuf));
+            _parent->_num_batches++;
+            bool max_bytes_reached = _parent->_buffer.size_bytes()
+                                     > static_cast<size_t>(_parent->_max_bytes);
```

Comment:
Is the expectation that `max_bytes` will roughly limit the size of a remote segment? Or is it only exposed as a way to bound the buffer size?